week-4-tests-and-persistence/w04e02-test-nest/src/users/auth/auth.service.spec.ts

```diff
@@ -15,4 +17,20 @@ describe('AuthService', () => {
   it('should be defined', () => {
     expect(service).toBeDefined();
   });
+
+  it('should throw error when wrong credentials', async () => {
+    await expect(
+      service.signIn('jakub@example.pl', 'pass123'),
+    ).rejects.toThrowError();
+  });
+
+  it('should return token when correct credentials', async () => {
+    const token = await service.signIn('peter@myclient.com', 'som3pa55w0rd');
+    expect(token).toBeDefined();
```

Comment:
Tutaj zaczyna sie prawdziwa "zabawa".

Przede wszystkim żeby przetestować logikę `signIn` musisz zaFakeować metodę:
```
 getOne() { }
```
na fikcyjnym `User service`

Musi ona zwracać prawidłowy z perspektywy systemu `hash` hasła + `id` usera minimum, żeby móc lecieć dalej.

W tym układzie musimy zahashować @teziovsky za wczasu hasło przez `bcrypt` żeby mógł on je sprawdzić.

Zobacz:

```typescript
providers: [
        AuthService,
        {
          provide: UsersService,
          useValue: {
            getOne() {
              // Fake record (niby z DB)
              return {
                id: 10,
                password: validPass,
              };
            },
            saveToken() {},
          },
        },
      ],
```

Gdzie `validPass` to:

```typescript
 const validPass = await bcrypt.hash('boom', 8);
```

Dodatkowo mamy tam Stub `saveToken` ponieważ metoda `.signIn` używa tego pod spodem.
Nie musimy tym nic zwracać.

Ale co Ciekawe, nasz `authToken` jest zwracany na podstawie akutalnej daty.
Możemy więc w `jest` zamockować jaka jest data systemowa, i podstawić stałą :)

Więc cały test wygląda tak:

```typescript
it('should return valid token based on correct date when correct credentials', async () => {
    jest.useFakeTimers().setSystemTime(new Date('2022-12-06'));

    const token = await service.signIn('peter@myclient.com', 'boom');

    expect(token).toEqual('lbbgg00a');
  });
```

Zwróć uwagę na hasło `boom` - można zrobić osobny test sprawdzający czy jest używany `bcrypt`. I dać go wcześniej. Wtedy nie interesuje nas token, i możemy dać `.toBeDefined`

A tutaj testujemy czy faktycznie `token` jest `valid` :).